File: 30-client-sdk/32-client-identify/react/flag-controls.js
/**
 * Lab Controls proxy — LaunchDarkly REST from Node, never from the page.
 *
 * Feature flags PATCH (semantic patch)
 * https://launchdarkly.com/docs/api/feature-flags/patch-feature-flag
 * Keywords: turnFlagOn, turnFlagOff, updateFallthroughVariationOrRollout
 */

const FLAG_HIGHLIGHT = "enable-identify-grid-highlight";
const FLAG_COUNT = "show-identify-move-count";
const DEFAULT_ON_COLOR = "green";

const CONTROLLED = [
  {
    key: FLAG_HIGHLIGHT,
    label: "Identify grid highlight",
    summary:
      "String flag with key targeting: alice→green, bob→blue, else none. Leave ON for the identify demo.",
  },
  {
    key: FLAG_COUNT,
    label: "Identify move count",
    summary:
      "Boolean flag with key targeting: alice→visible, bob/else hidden. Leave ON for the identify demo.",
  },
];

const ALLOWED = new Set(CONTROLLED.map((item) => item.key));

function apiConfig() {
  const token = (process.env.LD_API_ACCESS_TOKEN || "").trim();
  const project = (process.env.LD_PROJECT_KEY || "").trim();
  const environment = (process.env.LD_ENVIRONMENT_KEY || "").trim();
  const missing = [];
  if (!token) missing.push("LD_API_ACCESS_TOKEN");
  if (!project) missing.push("LD_PROJECT_KEY");
  if (!environment) missing.push("LD_ENVIRONMENT_KEY");
  return {
    configured: missing.length === 0,
    missing,
    projectKey: project || null,
    environmentKey: environment || null,
    apiHost: process.env.LD_API_HOST || "https://app.launchdarkly.com",
  };
}

function ldRequest(method, path, body) {
  const cfg = apiConfig();
  if (!cfg.configured) {
    const err = new Error("Flag controls need " + cfg.missing.join(", ") + " in the server environment.");
    err.status = 503;
    throw err;
  }
  const token = (process.env.LD_API_ACCESS_TOKEN || "").trim();
  const version = process.env.LD_API_VERSION || "20240415";
  const url = `${cfg.apiHost.replace(/\/$/, "")}/api/v2${path}`;
  const headers = {
    Authorization: token,
    "LD-API-Version": version,
    Accept: "application/json",
  };
  const opts = { method, headers };
  if (body !== undefined) {
    headers["Content-Type"] =
      method === "PATCH"
        ? "application/json; domain-model=launchdarkly.semanticpatch"
        : "application/json";
    opts.body = JSON.stringify(body);
  }
  return fetch(url, opts).then(async (res) => {
    const text = await res.text();
    const parsed = text ? JSON.parse(text) : {};
    if (!res.ok) {
      const err = new Error(parsed.message || text || `LaunchDarkly API ${res.status}`);
      err.status = res.status;
      throw err;
    }
    return parsed;
  });
}

function variationValue(flag, index) {
  const variations = flag.variations || [];
  if (index == null || index < 0 || index >= variations.length) return null;
  return variations[index].value;
}

function variationIdForValue(flag, wanted) {
  const want = typeof wanted === "string" ? wanted.trim().toLowerCase() : wanted;
  for (const variation of flag.variations || []) {
    const val = variation.value;
    const match =
      val === wanted ||
      (typeof val === "string" && typeof want === "string" && val.trim().toLowerCase() === want);
    if (!match) continue;
    const id = variation._id || variation.id;
    if (typeof id === "string" && id) return id;
  }
  return null;
}

function isOffValue(val) {
  return val === "none" || val === false;
}

function summarizeFlag(flag, environmentKey, meta) {
  const env = (flag.environments || {})[environmentKey] || {};
  const on = Boolean(env.on);
  const offIdx = env.offVariation;
  const fallIdx = (env.fallthrough || {}).variation;
  const offValue = variationValue(flag, typeof offIdx === "number" ? offIdx : null);
  const fallValue = variationValue(flag, typeof fallIdx === "number" ? fallIdx : null);
  const key = flag.key || meta.key;
  const stringVars = (flag.variations || []).every((v) => typeof v.value === "string");
  const colorOptions = [];
  if (key === FLAG_HIGHLIGHT && stringVars) {
    for (const variation of flag.variations || []) {
      if (typeof variation.value === "string" && !isOffValue(variation.value)) {
        colorOptions.push(variation.value);
      }
    }
  }
  let targetingHint;
  if (on) {
    targetingHint = `Flag is ON — fallthrough ${JSON.stringify(fallValue)}.`;
  } else {
    targetingHint = `Flag is OFF — off variation ${JSON.stringify(offValue)}.`;
  }
  return {
    key,
    name: flag.name || meta.label,
    label: meta.label,
    summary: meta.summary,
    on,
    variationKind: key === FLAG_HIGHLIGHT ? "string" : "boolean",
    colorOptions,
    servedWhenOff: offValue,
    servedWhenOnFallthrough: fallValue,
    targetingHint,
  };
}

function fallthroughInstruction(flag, preferred) {
  if (flag.key !== FLAG_HIGHLIGHT) return null;
  const color = (preferred || DEFAULT_ON_COLOR).trim().toLowerCase();
  let colorId = variationIdForValue(flag, color);
  if (!colorId) {
    for (const variation of flag.variations || []) {
      if (typeof variation.value === "string" && !isOffValue(variation.value)) {
        colorId = variation._id || variation.id;
        break;
      }
    }
  }
  if (!colorId) return null;
  return { kind: "updateFallthroughVariationOrRollout", variationId: colorId };
}

async function listFlagControls() {
  const cfg = apiConfig();
  if (!cfg.configured) {
    return {
      ...cfg,
      flags: CONTROLLED.map((item) => ({
        ...item,
        on: null,
        targetingHint: "Set missing env vars to enable controls.",
      })),
    };
  }
  const query = new URLSearchParams({ env: cfg.environmentKey }).toString();
  const flags = [];
  const errors = [];
  for (const meta of CONTROLLED) {
    try {
      const flag = await ldRequest("GET", `/flags/${cfg.projectKey}/${meta.key}?${query}`);
      flags.push(summarizeFlag(flag, cfg.environmentKey, meta));
    } catch (exc) {
      errors.push({ key: meta.key, error: String(exc.message || exc) });
      flags.push({
        ...meta,
        on: null,
        targetingHint: String(exc.message || exc),
        error: String(exc.message || exc),
      });
    }
  }
  return { ...cfg, flags, errors };
}

async function applyFlagControl(flagKey, { on, fallthrough } = {}) {
  if (!ALLOWED.has(flagKey)) {
    const err = new Error(`Flag key not allowed for controls: ${flagKey}`);
    err.status = 400;
    throw err;
  }
  if (on === undefined && fallthrough == null) {
    const err = new Error('Provide "on" and/or "fallthrough"');
    err.status = 400;
    throw err;
  }
  const cfg = apiConfig();
  const query = new URLSearchParams({ env: cfg.environmentKey }).toString();
  let flag = await ldRequest("GET", `/flags/${cfg.projectKey}/${flagKey}?${query}`);
  const instructions = [];
  if (on === true) {
    instructions.push({ kind: "turnFlagOn" });
    const fall = fallthroughInstruction(flag, fallthrough);
    if (fall) instructions.push(fall);
  } else if (on === false) {
    instructions.push({ kind: "turnFlagOff" });
    if (fallthrough != null) {
      const fall = fallthroughInstruction(flag, fallthrough);
      if (fall) instructions.push(fall);
    }
  } else if (fallthrough != null) {
    const fall = fallthroughInstruction(flag, fallthrough);
    if (!fall) {
      const err = new Error(`No string variation matching fallthrough=${fallthrough}`);
      err.status = 400;
      throw err;
    }
    instructions.push(fall);
  }
  await ldRequest("PATCH", `/flags/${cfg.projectKey}/${flagKey}`, {
    environmentKey: cfg.environmentKey,
    comment: "32-client-identify UI",
    instructions,
  });
  flag = await ldRequest("GET", `/flags/${cfg.projectKey}/${flagKey}?${query}`);
  const meta = CONTROLLED.find((item) => item.key === flagKey);
  return {
    ok: true,
    projectKey: cfg.projectKey,
    environmentKey: cfg.environmentKey,
    flag: summarizeFlag(flag, cfg.environmentKey, meta),
  };
}

export {
  FLAG_HIGHLIGHT,
  FLAG_COUNT,
  apiConfig,
  listFlagControls,
  applyFlagControl,
};

File: 10-code-control/14-multi-context-targeting/dotnet/FlagControls.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json.Nodes;

/// <summary>
/// Controls flag on/off and fallthrough without editing provisioned targeting rules.
/// https://launchdarkly.com/docs/api/feature-flags/patch-feature-flag
/// </summary>
sealed class FlagControls
{
    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public object ApiConfig() => new
    {
        configured = Missing().Count == 0,
        missing = Missing(),
        projectKey = Env("LD_PROJECT_KEY"),
        environmentKey = Env("LD_ENVIRONMENT_KEY"),
        apiHost = ApiHost,
    };

    public async Task<object> ListAsync()
    {
        if (Missing().Count > 0)
            return ConfigWithFlags([
                new
                {
                    key = Partner.FlagKey,
                    label = "Show: partner org badge",
                    summary = "Boolean targeted by user+organization multi-context rules.",
                    on = (bool?)null,
                    targetingHint = "Set missing environment variables to enable controls.",
                },
            ]);
        var flag = await GetFlagAsync();
        return ConfigWithFlags([Summarize(flag)], []);
    }

    public async Task<object> ApplyAsync(
        string key, bool? on, bool hasFallthrough, JsonNode? fallthrough)
    {
        if (key != Partner.FlagKey)
            throw new ArgumentException($"Flag key not allowed for controls: {key}");
        if (on is null && !hasFallthrough)
            throw new ArgumentException("Provide \"on\" and/or \"fallthrough\"");
        RequireConfigured();

        var flag = await GetFlagAsync();
        var instructions = new JsonArray();
        if (on is true) instructions.Add(new JsonObject { ["kind"] = "turnFlagOn" });
        else if (on is false) instructions.Add(new JsonObject { ["kind"] = "turnFlagOff" });
        if (hasFallthrough)
        {
            var variationId = VariationId(flag, fallthrough);
            if (variationId is null)
                throw new ArgumentException($"No variation matching fallthrough={fallthrough?.ToJsonString()}");
            instructions.Add(new JsonObject
            {
                ["kind"] = "updateFallthroughVariationOrRollout",
                ["variationId"] = variationId,
            });
        }
        await RequestAsync(HttpMethod.Patch,
            $"/flags/{Escape(Project)}/{Partner.FlagKey}",
            new JsonObject
            {
                ["environmentKey"] = EnvironmentKey,
                ["comment"] = "14-multi-context-targeting UI: on/off or fallthrough",
                ["instructions"] = instructions,
            });
        flag = await GetFlagAsync();
        return new
        {
            ok = true,
            action = string.Join("+", instructions.Select(item => item?["kind"]?.ToString())),
            instructions = instructions.Select(item => item?["kind"]?.ToString()).ToArray(),
            projectKey = Project,
            environmentKey = EnvironmentKey,
            flag = Summarize(flag),
        };
    }

    private object ConfigWithFlags(List<object> flags, List<object>? errors = null) => new
    {
        configured = Missing().Count == 0,
        missing = Missing(),
        projectKey = Env("LD_PROJECT_KEY"),
        environmentKey = Env("LD_ENVIRONMENT_KEY"),
        apiHost = ApiHost,
        flags,
        errors = errors ?? [],
    };

    private static object Summarize(JsonObject flag)
    {
        var environment = flag["environments"]?[EnvironmentKey] as JsonObject ?? new JsonObject();
        var variations = flag["variations"] as JsonArray ?? [];
        var fallIndex = Int(environment["fallthrough"]?["variation"]);
        var offIndex = Int(environment["offVariation"]);
        var fallValue = VariationValue(variations, fallIndex);
        var options = variations.Select(node => new
        {
            token = OptionToken(node?["value"]),
            label = node?["name"]?.ToString() ?? OptionToken(node?["value"]),
            value = node?["value"],
        }).ToArray();
        var ruleCount = (environment["rules"] as JsonArray)?.Count ?? 0;
        return new
        {
            key = Partner.FlagKey,
            name = flag["name"]?.ToString() ?? "Show partner org badge",
            label = "Show: partner org badge",
            summary = "Boolean true only for alice+acme and bob+globex (provisioned AND rules).",
            on = environment["on"]?.GetValue<bool>() ?? false,
            variationKind = "boolean",
            fallthroughOptions = options,
            fallthroughToken = OptionToken(JsonValueFrom(fallValue)),
            servedWhenOff = VariationValue(variations, offIndex),
            servedWhenOnFallthrough = fallValue,
            ruleCount,
            targetingHint = $"{ruleCount} provisioned multi-context rules remain unchanged; this lab controls only flag state and fallthrough.",
        };
    }

    private async Task<JsonObject> GetFlagAsync() =>
        await RequestAsync(HttpMethod.Get,
            $"/flags/{Escape(Project)}/{Partner.FlagKey}?env={Escape(EnvironmentKey)}");

    private async Task<JsonObject> RequestAsync(
        HttpMethod method, string path, JsonObject? body = null)
    {
        RequireConfigured();
        using var request = new HttpRequestMessage(method, $"{ApiHost.TrimEnd('/')}/api/v2{path}");
        request.Headers.TryAddWithoutValidation("Authorization", Token);
        request.Headers.Add("LD-API-Version", Env("LD_API_VERSION") ?? "20240415");
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        if (body is not null)
        {
            request.Content = new StringContent(body.ToJsonString(), Encoding.UTF8);
            request.Content.Headers.ContentType =
                MediaTypeHeaderValue.Parse("application/json; domain-model=launchdarkly.semanticpatch");
        }
        using var response = await _http.SendAsync(request);
        var raw = await response.Content.ReadAsStringAsync();
        if (!response.IsSuccessStatusCode)
        {
            var message = JsonNode.Parse(raw)?["message"]?.ToString() ?? raw;
            throw new InvalidOperationException($"LaunchDarkly API {(int)response.StatusCode}: {message}");
        }
        return string.IsNullOrWhiteSpace(raw)
            ? new JsonObject()
            : JsonNode.Parse(raw)?.AsObject() ?? new JsonObject();
    }

    private static string? VariationId(JsonObject flag, JsonNode? wanted)
    {
        var normalized = NormalizeWanted(wanted);
        foreach (var node in flag["variations"] as JsonArray ?? [])
        {
            if (ValuesEqual(node?["value"], normalized))
                return node?["_id"]?.ToString() ?? node?["id"]?.ToString();
        }
        return null;
    }

    private static object? VariationValue(JsonArray values, int? index)
    {
        if (index is not >= 0 || index >= values.Count) return null;
        var node = values[index.Value]?["value"];
        if (node is JsonValue value)
        {
            if (value.TryGetValue<bool>(out var flag)) return flag;
            if (value.TryGetValue<string>(out var text)) return text;
            if (value.TryGetValue<int>(out var number)) return number;
        }
        return node;
    }

    private static JsonNode? JsonValueFrom(object? value) => value switch
    {
        bool flag => JsonValue.Create(flag),
        string text => JsonValue.Create(text),
        int number => JsonValue.Create(number),
        _ => null,
    };

    private static string OptionToken(JsonNode? node)
    {
        if (node is JsonValue value)
        {
            if (value.TryGetValue<bool>(out var flag)) return flag ? "true" : "false";
            if (value.TryGetValue<string>(out var text)) return text;
        }
        return node?.ToString() ?? "";
    }

    private static object? NormalizeWanted(JsonNode? wanted)
    {
        if (wanted is JsonValue value)
        {
            if (value.TryGetValue<bool>(out var flag)) return flag;
            if (value.TryGetValue<string>(out var text))
            {
                if (text == "true") return true;
                if (text == "false") return false;
                return text;
            }
        }
        return wanted?.ToString();
    }

    private static bool ValuesEqual(JsonNode? left, object? wanted)
    {
        if (left is JsonValue value)
        {
            if (wanted is bool flag && value.TryGetValue<bool>(out var leftFlag))
                return leftFlag == flag;
            if (wanted is string text && value.TryGetValue<string>(out var leftText))
                return leftText == text;
        }
        return false;
    }

    private static int? Int(JsonNode? node) =>
        node is JsonValue value && value.TryGetValue<int>(out var number) ? number : null;
    private static string Escape(string value) => Uri.EscapeDataString(value);
    private static string? Env(string key) =>
        Environment.GetEnvironmentVariable(key)?.Trim() is { Length: > 0 } value ? value : null;
    private static List<string> Missing() =>
        new[] { "LD_API_ACCESS_TOKEN", "LD_PROJECT_KEY", "LD_ENVIRONMENT_KEY" }
            .Where(key => Env(key) is null).ToList();
    private static void RequireConfigured()
    {
        var missing = Missing();
        if (missing.Count > 0)
            throw new InvalidOperationException($"Flag controls need {string.Join(", ", missing)}");
    }
    private static string Token => Env("LD_API_ACCESS_TOKEN")!;
    private static string Project => Env("LD_PROJECT_KEY")!;
    private static string EnvironmentKey => Env("LD_ENVIRONMENT_KEY")!;
    private static string ApiHost => Env("LD_API_HOST") ?? "https://app.launchdarkly.com";
}
